### ISSUE TITLE
ci: modernize CI pipeline — replace deprecated actions-rs, enforce clippy -D warnings, add tarpaulin coverage, rustfmt, and pre-commit hooks

### ISSUE THREAD

**Issue by e-desouza** [sephynox/xrpl-rust#139]
## Summary

Several improvements are needed to harden the CI pipeline and add local developer tooling for code quality enforcement.

---

### 1. Deprecated `actions-rs` actions

All three existing workflows use actions from the `actions-rs` GitHub organization, which is **archived and no longer maintained**:

- `actions-rs/clippy-check@v1` (quality_test.yml)
- `actions-rs/cargo@v1` and `actions-rs/toolchain@v1` (unit_test.yml)
- `actions-rs/audit-check@v1` (audit_test.yml)

These should be replaced with direct `cargo` invocations via `run:` steps or actively maintained alternatives like `dtolnay/rust-toolchain`.

### 2. Clippy does not treat warnings as errors

The current `clippy_check` step runs `--all-features` but does not pass `-D warnings`, meaning clippy warnings are reported but do not fail the build. This allows linting debt to accumulate silently.

### 3. No code coverage reporting

There is no coverage step in any workflow. Contributors have no visibility into which code paths are tested.

### 4. No `rustfmt` enforcement

There is no CI check for `cargo fmt -- --check`, so formatting inconsistencies can be merged without notice.

### 5. No pre-commit hooks

There are no local pre-commit hooks to catch issues (fmt, clippy, test) before a push. Contributors have to rely entirely on CI feedback cycles.

### 6. `quality_test.yml` does not trigger on PRs to `dev`

The quality workflow triggers on `push: dev` and `pull_request: main`, but not on `pull_request: dev`, so clippy feedback is missing for feature branches targeting `dev`.

---

#### Replace `actions-rs` with `dtolnay/rust-toolchain` + direct `cargo` steps

```yaml
- uses: dtolnay/rust-toolchain@stable
  with:
    components: clippy, rustfmt
```

#### Enforce clippy as a hard failure

```yaml
- name: Clippy (deny warnings)
  run: cargo clippy --all-features -- -D warnings
```

#### Add `rustfmt` check

```yaml
- name: Format check
  run: cargo fmt --all -- --check
```

#### Add code coverage via `cargo-tarpaulin`

```yaml
- name: Install tarpaulin
  run: cargo install cargo-tarpaulin

- name: Coverage
  run: cargo tarpaulin --all-features --workspace --timeout 120 --out Xml

- name: Upload coverage to Codecov
  uses: codecov/codecov-action@v4
  with:
    files: cobertura.xml
    fail_ci_if_error: false
```

> **Alternative**: use `cargo-llvm-cov` if tarpaulin has compatibility issues with `no_std` feature combinations.

#### Fix `quality_test.yml` trigger to include PRs to `dev`

```yaml
on:
  push:
    branches: [dev]
  pull_request:
    branches: [main, dev]
```

---

### Pre-Commit Hooks

Add a `.pre-commit-config.yaml` using the [pre-commit](https://pre-commit.com) framework:

```yaml
repos:
  - repo: local
    hooks:
      - id: cargo-fmt
        name: cargo fmt
        entry: cargo fmt --all --
        language: system
        types: [rust]
        pass_filenames: false

      - id: cargo-clippy
        name: cargo clippy
        entry: cargo clippy --all-features -- -D warnings
        language: system
        types: [rust]
        pass_filenames: false

      - id: cargo-test
        name: cargo test
        entry: cargo test --release
        language: system
        types: [rust]
        pass_filenames: false
```

Contributors install hooks with:

```bash
pip install pre-commit
pre-commit install
```

---

## Acceptance Criteria

- [ ] All `actions-rs` usages replaced with `dtolnay/rust-toolchain` or direct `cargo` steps
- [ ] `cargo clippy -- -D warnings` enforced across all feature combinations in CI
- [ ] `cargo fmt --all -- --check` added to `quality_test.yml`
- [ ] Code coverage measured via `cargo-tarpaulin` (or `cargo-llvm-cov`) and reported to Codecov
- [ ] `quality_test.yml` triggers on PRs to both `main` and `dev`
- [ ] `.pre-commit-config.yaml` added with fmt, clippy, and test hooks
- [ ] `CONTRIBUTING.md` or `README.md` updated with pre-commit setup instructions
